archive/100-theorems-list/pnat_perfect.lean

```diff
@@ -0,0 +1,509 @@
+/-
+Copyright (c) 2020 Aaron Anderson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Aaron Anderson.
+-/
+
+import data.nat.totient
+import number_theory.quadratic_reciprocity
+import number_theory.lucas_lehmer
+import multiplicity_vectors
+
+
+/-!
+# Perfect Numbers
+
+We define perfect numbers and prove the Euclid-Euler theorem relating them to Mersenne primes
+
+## Notations
+
+## Implementation Notes
+I have used pnats in this version most of the time.
+As a result, there are many new pnat facts and definitions in multiplicity_vectors.lean
+
+## References
+https://en.wikipedia.org/wiki/Euclid%E2%80%93Euler_theorem
+-/
+
+open_locale classical
+open_locale big_operators
+
+section definitions
+
+variable (n : ℕ+)
+
+/--
+The finset of (positive) divisors of n
+-/
+def divisors : finset ℕ+ := (pnat.Ico 1 (n + 1)).filter (λ x : ℕ+, x ∣ n)
+
+/--
+The finset of proper (positive) divisors of n
+-/
+def proper_divisors : finset ℕ+ := (pnat.Ico 1 n).filter (λ x : ℕ+, x ∣ n)
+
+lemma not_proper_self : ¬ n ∈ proper_divisors n := by simp [proper_divisors]
+
+variable {n}
+lemma pnat.le_of_dvd {m : ℕ+} : m ∣ n → m ≤ n :=
+by { rw pnat.dvd_iff', intro h, rw ← h, apply (pnat.mod_le n m).left }
+
+@[simp]
+lemma mem_divisors {m : ℕ+} : m ∈ divisors n ↔ m ∣ n :=
+begin
+  rw divisors,
+  simp only [true_and, pnat.one_le, pnat.Ico.mem, finset.mem_filter],
+  split, intro hyp, exact hyp.right,
+  intro hyp, split, swap, exact hyp,
+  rw pnat.lt_add_one_iff, apply pnat.le_of_dvd hyp
+end
+
+lemma divisor_le {m : ℕ+}:
+m ∈ divisors n → m ≤ n := by {rw mem_divisors, exact pnat.le_of_dvd}
+
+@[simp]
+lemma mem_proper_divisors {m : ℕ+} : m ∈ proper_divisors n ↔ m ∣ n ∧ m ≠ n :=
+begin
+  rw proper_divisors,
+  simp only [true_and, pnat.one_le, ne.def, pnat.Ico.mem, finset.mem_filter],
+  split, intro hyp, split, exact hyp.right,
+  { intro contra, rw contra at hyp, apply lt_irrefl n hyp.left },
+  { intro hyp, split, swap, exact hyp.left, apply lt_of_le_of_ne,
+    apply pnat.le_of_dvd hyp.left, apply hyp.right }
+end
+
+variable (n)
+
+----pnat.dvd_refl should be a simp lemma!
+lemma divisors_eq_proper_divisors_insert_self :
+  divisors n = has_insert.insert n (proper_divisors n) :=
+begin
+  ext, rw [finset.mem_insert, mem_divisors, mem_proper_divisors], split,
+  { intro hdvd, by_cases a = n, left, exact h,
+    right, split, exact hdvd, exact h },
+  { intro hyp, cases hyp, rw hyp, apply pnat.dvd_refl, apply hyp.left }
+end
+
+/--The sum (nat) of the divisors of a pnat n-/
+def sum_divisors : ℕ := ∑ i in divisors n, i
```

Comment:
Why don't you go for `pnat` directly? Every new definition needs 10 supporting simp-lemmas and connections with all other relevant definitions.